### DISCUSSION TITLE
Integration of Large Language Models (LLMs) with Ghidra

### DISCUSSION THREAD

**Discussion by KatoStevenMubiru** [NationalSecurityAgency/ghidra#7088] [Ideas]
**Description:**
Propose integrating LLMs (e.g., OpenAI's GPT) to enhance Ghidra's analysis capabilities by providing natural language explanations of complex assembly code and automating the generation of documentation.

**Benefits:**
- **Enhanced Understanding:** Helps analysts quickly grasp intricate code structures and functionalities.
- **Automation:** Streamlines the documentation process, saving time and reducing manual effort.
- **Accessibility:** Makes reverse engineering more accessible to those less familiar with assembly languages.

**Proposed Implementation:**
- **Plugin Development:** Create a Ghidra plugin that interfaces with an LLM API to process selected code segments.
- **Natural Language Output:** Display LLM-generated explanations within Ghidra's UI.
- **Customization Options:** Allow users to configure the depth and style of the explanations.

**Additional Context:**
Integrating LLMs can bridge the gap between low-level code analysis and high-level understanding, making Ghidra an even more powerful tool for both novice and experienced analysts.

**Comment by KatoStevenMubiru** [comment]
Hello @everyone , I would want to know from the community, why this idea has got many down votes so far?
May be I could refine it abit

**Comment by hippietrail** [comment]
I would say just forge ahead and start making such a thing as a Ghidra extension on your own. Ask for help with particular things you get stuck with along the way. Maybe your extension will need changes to Ghidra itself.

I use LLMs all the time when I'm coding, but I've never tried to interface with an LLM programmatically so I have little clue where to start. This will work best if you have a very clear vision of where you would start.

In the meantime you can definitely already copy and paste from the listing or the decompiler into an LLM site and ask how the code works, etc.

So far LLMs are based on text so at present they wouldn't be able to help resolving jump tables or other things that would involve analysing binary bytes rather than already disassembled code. In the future they will probably be able to find exploits and see through obfuscation, follow values passing through registers and memory addresses, etc.

Probably the most obvious and straightforward place to start would be an "add comments" function.

**Reply by xxxajk** [reply]
That's funny, the llm model I use can indeed understand tables, and is from 2023. use a model that has a high SWE score, not a silly generic one, because those win indeed fall flat.
and yes, i know this is an old thread... I am correcting the misunderstandings I find.

**Comment by KatoStevenMubiru** [comment]
@hippietrail

Thank you for the encouragement and helpful suggestions! I'll start developing the Ghidra extension independently to integrate LLMs, beginning with the "add comments" feature. I'll reach out if I need any assistance along the way.

Looking forward to contributing!

**Comment by thixotropist** [comment]
You might take a look at discussion #6045, which speculates that your kind of work might apply to making sense of binaries compiled with vector extensions supported.  The GCC test suite provides a decent training set of C source code compiled to vector instruction sequences for different machine architectures.

Optimizing compilers can turn simple loops over arrays of structures into code sequences Ghidra struggles with.

**Comment by hippietrail** [comment]
I just had an idea. Maybe the next easiest AI feature to implement might be naming variables and functions, and more useful than adding comments too.

**Reply by Wall-AF** [reply]
> I just had an idea. Maybe the next easiest AI feature to implement might be naming variables and functions, and more useful than adding comments too.

There is an oustanding feature request that, if implemented, would go along way towards a better automated naming process, #6111.

**Comment by KatoStevenMubiru** [comment]
Thank you @thixotropist  and @hippietrail . Great insights so far.

**Comment by thixotropist** [comment]
You likely need some good examples of what you would like your tool to generate.  I like using [Whisper.cpp](https://github.com/ggerganov/whisper.cpp) as the C and C++ source built with the latest GCC compiler with vector extensions enabled and various machine architecture settings.  Your tool might start by analyzing the Ghidra assembly and decompiler windows to say 'that looks like an inlined vector dot product'.  The trick is to go further and have it say '... but this implementation includes a serious buffer overrun/underrun error.'

**Comment by ghidra2** [comment]
There are certainly some cases where LLMs can be useful in the context of reverse engineering.  However, it is not always the case that the LLM will provide useful information which makes their integration into Ghidra contentious.  I would like to see more evidence of consistent and reliable results before we support such a capability.

**Reply by hippietrail** [reply]
Probably the number 1 reason to do it as an extension.

**Comment by KatoStevenMubiru** [comment]
Thanks @ghidra2  , point taken .

**Comment by nblog** [comment]
😅 I was one of those who initially voted against integrating LLMs, but after witnessing how rapidly AI technology has advanced, I realize now how shortsighted that decision was. Especially after trying out state-of-the-art models like `Google Gemini 2.5 Pro` and `Anthropic Claude Sonnet 4`, I’m convinced that leveraging LLMs can significantly reduce our workload.

I've already come across some amazing MCPs built on top of Ghidra that utilize these models, and their performance is truly impressive. I can only imagine how incredible the experience would be if LLMs were deeply integrated into Ghidra itself. It could really revolutionize the way we work!

**Reply by astrelsky** [reply]
Same. Unfortunately I have still never used any. I would need it to run locally for ghidra but they always seem complicated to setup and I can't get past that hurdle.

**Reply by nblog** [reply]
> Same. Unfortunately I have still never used any. I would need it to run locally for ghidra but they always seem complicated to setup and I can't get past that hurdle.

Ghidra: https://github.com/LaurieWired/GhidraMCP
IDA Pro: https://github.com/mrexodia/ida-pro-mcp
WinDbg: https://github.com/svnscha/mcp-windbg
All of these repositories come with excellent documentation, so you can quickly achieve similar results by following their instructions.

If you have a Claude API KEY, that's a great place to start. Alternatively, you can use VSCode Copilot. For other LLM platforms, the "Roo Code" extension for VSCode is also a solid choice.

Once you overcome the initial hurdles and get things running for the first time, you'll likely feel a real sense of excitement—especially when you use more advanced and powerful models, such as Gemini 2.5 Pro or Claude 4. The results they deliver can be truly impressive.

Let me share a quick example from my own experience: I opened my project directory in VSCode and, using a combination of Copilot and Claude 4 in agent mode alongside [windbg-ext-mcp](https://github.com/NadavLor/windbg-ext-mcp), I asked the AI to help me analyze a crash dump (dmp), identify the root cause, locate the error in my code, and even suggest code fixes. The whole process was seamless and incredibly impressive—I was genuinely amazed at the capabilities these tools brought to my workflow.

**Comment by xxxajk** [comment]
I question all of the efforts on using cloud crap. run locally, and run a model with a high SWE benchmark rating.
It will eat up ghidra output like candy and produce some very good analysis.

**Comment by xxxajk** [comment]
I will provide a simpler example of how good use of a properly run LLM with own-built CLI can do things that will surprise you, and yeah, Ghidra wasn't used. 
As a test, I provided unlabeled Z80 disassembly, and did not tell the LLM what the hand-optimized code was. The same worked if I gave it assembly or just presented it with the raw bytes, but told the LLM that it was Z80 code.
The only hint i had to tell the llm was to take the code, and turn it into the C equivalent.
The goal of the test was to see if it could identify what the purpose or function of the code was.
The local model ran was devstral-small-2:24b on the very affordable, but slow Nvidia P40, and the time it took to do the analysis was quite acceptable, being about 7 minutes. Best of all it was spot-on and accurate.
The code is hand optimized Z80 assembly that calculates IP checksum. The idea is that I know for a fact that code like this is NOT in any training data anywhere, because I wrote it over 20 years ago, and the source was never published.
Those facts ensured that the model had to work it out using reasoning and recognition.
Unfortunately I can't  publish the code, because we all know it would get swallowed up into training data, and that would ruin the test. I can (if requested) publish the output here, if there is any interest. :-)
